File: gatewayd/sdk/contract.go
// Package sdk is what a tool service embeds to join a Flory deployment.
//
// It is the only implementation of the gateway protocol on the service side: no
// tool service hand-writes registration, heartbeat, or health. That matters
// because those three are what make a stateless gateway recoverable, and a
// service that got any of them subtly wrong would look registered while being
// unroutable, or keep heartbeating at a gateway that has forgotten it.
//
// The SDK is domain-neutral. It knows about contracts, not about inventory.
package sdk

import (
	"errors"
	"fmt"

	gatewayv1 "github.com/linuxb/flory-ai/gatewayd/internal/pb/flory/gateway/v1"
	"github.com/linuxb/flory-ai/gatewayd/internal/registry"
	"github.com/linuxb/flory-ai/gatewayd/internal/toolview"
)

// Contract describes one tool a service offers, in the terms a service author
// thinks in. It is projected onto the wire ToolContract by Build.
//
// There is deliberately no field for IsPivot or Compensable: both are derived
// from EffectClass and the declared undo path (02 section 2.1), so there is
// nothing here that could disagree with what they are derived from.
type Contract struct {
	ToolID      string
	ToolVersion string
	Description string
	// InputSchema and OutputSchema are JSON Schema documents as text. Tool
	// payloads stay JSON Schema because that is what MCP inputSchema and the
	// planner consume.
	InputSchema  string
	OutputSchema string
	// RouteID is the stable logical route, not an address. It is part of the
	// contract's identity; which instances serve it is not.
	EffectClass         EffectClass
	Mode                Mode
	IdempotencyKeyPath  string
	IdempotentRetryable bool
	TryTimeoutSeconds   uint32
	ConfirmTool         string
	CancelTool          string
	CompensateTool      string
	StatusTool          string
	// Compensating marks a tool that reverses another's effect. It always
	// registers as delta-based: releasing exactly what the matching try added is
	// the only form that commutes with another branch's committed change, and the
	// gateway refuses snapshot restore outright.
	Compensating bool
	Footprint    []string
	Writes       []string
	TimeoutMS    uint32
	Retry        Retry
	Owner        string
}

// EffectClass is a tool's side-effect classification.
type EffectClass string

// The effect classes from design document 02.
const (
	EffectNone         EffectClass = "none"
	EffectBufferable   EffectClass = "bufferable"
	EffectReversible   EffectClass = "reversible"
	EffectIrreversible EffectClass = "irreversible"
)

// Mode is a tool's transaction integration.
type Mode string

// The transaction modes from design document 02.
const (
	ModePlain Mode = "plain"
	ModeTCC   Mode = "tcc"
	ModeSaga  Mode = "saga"
)

// Retry is the envelope an executor may not exceed for this tool.
type Retry struct {
	MaxAttempts      uint32
	InitialBackoffMS uint32
	// Multiplier is expressed in thousandths, so the whole contract stays
	// integer-valued and the tool-view digest has a canonical text form.
	MultiplierMilli uint32
	MaxBackoffMS    uint32
}

// DefaultRetry is a conservative envelope for a service that has no opinion.
func DefaultRetry() Retry {
	return Retry{MaxAttempts: 3, InitialBackoffMS: 100, MultiplierMilli: 2000, MaxBackoffMS: 5000}
}

var effectClasses = map[EffectClass]gatewayv1.EffectClass{
	EffectNone:         gatewayv1.EffectClass_EFFECT_CLASS_NONE,
	EffectBufferable:   gatewayv1.EffectClass_EFFECT_CLASS_BUFFERABLE,
	EffectReversible:   gatewayv1.EffectClass_EFFECT_CLASS_REVERSIBLE,
	EffectIrreversible: gatewayv1.EffectClass_EFFECT_CLASS_IRREVERSIBLE,
}

var modes = map[Mode]gatewayv1.ToolMode{
	ModePlain: gatewayv1.ToolMode_TOOL_MODE_PLAIN,
	ModeTCC:   gatewayv1.ToolMode_TOOL_MODE_TCC,
	ModeSaga:  gatewayv1.ToolMode_TOOL_MODE_SAGA,
}

// Build projects a declared contract onto the wire form, for one route.
func (contract Contract) Build(routeID string) (*gatewayv1.ToolContract, error) {
	effectClass, known := effectClasses[contract.EffectClass]
	if !known {
		return nil, fmt.Errorf("sdk: %s declares an unknown effect class %q", contract.ToolID, contract.EffectClass)
	}
	mode, known := modes[contract.Mode]
	if !known {
		return nil, fmt.Errorf("sdk: %s declares an unknown mode %q", contract.ToolID, contract.Mode)
	}
	compensation := gatewayv1.CompensationStyle_COMPENSATION_STYLE_NOT_COMPENSATING
	if contract.Compensating {
		compensation = gatewayv1.CompensationStyle_COMPENSATION_STYLE_DELTA
	}
	retry := contract.Retry
	if retry.MaxAttempts == 0 {
		retry = DefaultRetry()
	}
	idempotent := contract.IdempotentRetryable
	return &gatewayv1.ToolContract{
		ToolId:       contract.ToolID,
		ToolVersion:  contract.ToolVersion,
		Description:  contract.Description,
		InputSchema:  contract.InputSchema,
		OutputSchema: contract.OutputSchema,
		RouteId:      routeID,
		Adapter:      &gatewayv1.AdapterSpec{Protocol: "grpc", Operation: "flory.gateway.v1.ToolExecutionService/Execute"},
		Txn: &gatewayv1.TransactionSpec{
			EffectClass:         effectClass,
			Mode:                mode,
			IdempotencyKeyPath:  contract.IdempotencyKeyPath,
			IdempotentRetryable: &idempotent,
			TryTimeoutS:         contract.TryTimeoutSeconds,
			ConfirmTool:         contract.ConfirmTool,
			CancelTool:          contract.CancelTool,
			CompensateTool:      contract.CompensateTool,
			StatusTool:          contract.StatusTool,
		},
		CompensationStyle: compensation,
		Footprint:         contract.Footprint,
		Writes:            contract.Writes,
		TimeoutMs:         contract.TimeoutMS,
		RetryConstraints: &gatewayv1.RetryConstraints{
			MaxAttempts:      retry.MaxAttempts,
			InitialBackoffMs: retry.InitialBackoffMS,
			MultiplierMilli:  retry.MultiplierMilli,
			MaxBackoffMs:     retry.MaxBackoffMS,
		},
		Owner: contract.Owner,
	}, nil
}

// Metadata returns the flory_transaction extension a listing carries for this
// contract, which is what the planner reads to derive transaction structure.
func Metadata(contract *gatewayv1.ToolContract) (map[string]any, error) {
	tool, err := toolview.FromProto(contract)
	if err != nil {
		return nil, err
	}
	return tool.Metadata(), nil
}

// validateContract runs the gateway's own structural rules against one contract.
//
// It calls the gateway's implementation rather than restating it. A second copy
// of these rules would eventually disagree with the first, and a service would
// discover its contract was bad at the worst possible moment: after deploying,
// from a rejection it had already convinced itself could not happen.
func validateContract(contract *gatewayv1.ToolContract) error {
	tool, err := toolview.FromProto(contract)
	if err != nil {
		return err
	}
	if violation := registry.ValidateStructure(tool); violation != nil {
		return errors.New(violation.Error())
	}
	return nil
}

File: sdk/typescript/contract.ts
import {create} from '@bufbuild/protobuf';
import {
    AdapterSpecSchema,
    CompensationStyle,
    EffectClass as WireEffectClass,
    RetryConstraintsSchema,
    type ToolContract,
    ToolContractSchema,
    ToolMode as WireToolMode,
    TransactionSpecSchema,
} from './gen/flory/gateway/v1/tool_contract_pb.js';

/** Side-effect classification from design document 02. */
export type EffectClass = 'none' | 'bufferable' | 'reversible' | 'irreversible';
/** Transaction integration supported by a registered tool. */
export type Mode = 'plain' | 'tcc' | 'saga';

/** The retry envelope an executor may not exceed for a tool. */
export interface Retry {
    maxAttempts: number;
    initialBackoffMs: number;
    /** Thousandths, so the whole contract stays integer-valued and the tool-view digest has a canonical text form. */
    multiplierMilli: number;
    maxBackoffMs: number;
}

/** A conservative envelope for a service that has no opinion. */
export function defaultRetry(): Retry {
    return {maxAttempts: 3, initialBackoffMs: 100, multiplierMilli: 2000, maxBackoffMs: 5000};
}

/**
 * One tool a service offers, in the terms a service author thinks in.
 *
 * There is deliberately no field for `isPivot` or `compensable`: both are derived from `effectClass` and the declared
 * undo path (02 section 2.1), so there is nothing here that could disagree with what they are derived from.
 */
export interface Contract {
    toolId: string;
    toolVersion: string;
    description?: string;
    /** JSON Schema text. Tool payloads stay JSON Schema because that is what MCP inputSchema and the planner consume. */
    inputSchema: string;
    outputSchema: string;
    effectClass: EffectClass;
    mode: Mode;
    idempotencyKeyPath?: string;
    idempotentRetryable: boolean;
    tryTimeoutSeconds?: number;
    confirmTool?: string;
    cancelTool?: string;
    compensateTool?: string;
    statusTool?: string;
    /**
     * Marks a tool that reverses another's effect. It always registers as delta-based: releasing exactly what the
     * matching try added is the only form that commutes with another branch's committed change, and the gateway
     * refuses snapshot restore outright.
     */
    compensating?: boolean;
    footprint?: readonly string[];
    writes?: readonly string[];
    timeoutMs: number;
    retry?: Retry;
    owner: string;
}

const EFFECT_CLASSES: Record<EffectClass, WireEffectClass> = {
    none: WireEffectClass.NONE,
    bufferable: WireEffectClass.BUFFERABLE,
    reversible: WireEffectClass.REVERSIBLE,
    irreversible: WireEffectClass.IRREVERSIBLE,
};

const MODES: Record<Mode, WireToolMode> = {
    plain: WireToolMode.PLAIN,
    tcc: WireToolMode.TCC,
    saga: WireToolMode.SAGA,
};

/**
 * Projects a declared contract onto the wire form for one route.
 *
 * The projection must match the Go SDK's byte for byte: a view containing tools from both languages has to have one
 * digest, so the language a service happens to be written in cannot be part of a contract's identity.
 */
export function buildContract(contract: Contract, routeId: string): ToolContract {
    const retry = contract.retry ?? defaultRetry();
    return create(ToolContractSchema, {
        toolId: contract.toolId,
        toolVersion: contract.toolVersion,
        description: contract.description ?? '',
        inputSchema: contract.inputSchema,
        outputSchema: contract.outputSchema,
        routeId,
        adapter: create(AdapterSpecSchema, {protocol: 'grpc', operation: 'flory.gateway.v1.ToolExecutionService/Execute'}),
        txn: create(TransactionSpecSchema, {
            effectClass: EFFECT_CLASSES[contract.effectClass],
            mode: MODES[contract.mode],
            idempotencyKeyPath: contract.idempotencyKeyPath ?? '',
            idempotentRetryable: contract.idempotentRetryable,
            tryTimeoutS: contract.tryTimeoutSeconds ?? 0,
            confirmTool: contract.confirmTool ?? '',
            cancelTool: contract.cancelTool ?? '',
            compensateTool: contract.compensateTool ?? '',
            statusTool: contract.statusTool ?? '',
        }),
        compensationStyle: contract.compensating ? CompensationStyle.DELTA : CompensationStyle.NOT_COMPENSATING,
        footprint: [...(contract.footprint ?? [])],
        writes: [...(contract.writes ?? [])],
        timeoutMs: contract.timeoutMs,
        retryConstraints: create(RetryConstraintsSchema, {
            maxAttempts: retry.maxAttempts,
            initialBackoffMs: retry.initialBackoffMs,
            multiplierMilli: retry.multiplierMilli,
            maxBackoffMs: retry.maxBackoffMs,
        }),
        owner: contract.owner,
    });
}

/**
 * Reports why a contract cannot be registered, or null if it can.
 *
 * This is a startup check against the rules the gateway will apply, so a service fails while someone is watching it
 * start rather than on a rejection they have to go read a log to find. The gateway remains the authority: this
 * mirrors only the rules decidable from one contract alone, and never reports a contract registrable that the gateway
 * would refuse.
 */
export function contractViolation(contract: Contract): string | null {
    for (const [field, value] of Object.entries({toolId: contract.toolId, toolVersion: contract.toolVersion, owner: contract.owner})) {
        if (!value.trim()) return `${field} is required`;
    }
    if (!schemaIsObject(contract.inputSchema)) return 'input_schema must be a JSON object';
    if (!schemaIsObject(contract.outputSchema)) return 'output_schema must be a JSON object';
    if (contract.timeoutMs <= 0) return 'timeout_ms is required';
    const retry = contract.retry ?? defaultRetry();
    if (retry.maxAttempts < 1) return 'retry.maxAttempts must be at least 1';
    if (retry.multiplierMilli < 1000) return 'retry.multiplierMilli must be at least 1000';
    if (contract.effectClass === 'none' && contract.mode !== 'plain') {
        return 'effect_class none has no side effect to bracket, so mode must be plain';
    }
    if (contract.effectClass === 'irreversible') {
        if (contract.mode !== 'plain') return 'effect_class irreversible has no undo path, so mode must be plain';
        if (contract.compensateTool || contract.confirmTool || contract.cancelTool) {
            return 'effect_class irreversible declares an undo path; a compensable effect is not irreversible';
        }
    }
    if (contract.effectClass === 'reversible' && contract.mode === 'plain' && !contract.compensateTool) {
        return 'effect_class reversible declares no undo path; a tool with no undo path must be registered irreversible';
    }
    if (contract.mode === 'tcc') {
        if (!contract.confirmTool || !contract.cancelTool) return 'mode tcc requires both confirmTool and cancelTool';
        if (!contract.tryTimeoutSeconds) return 'mode tcc requires a positive tryTimeoutSeconds';
    }
    if (contract.mode === 'saga' && !contract.compensateTool) return 'mode saga requires a compensateTool';
    for (const [field, companion] of Object.entries({
        confirmTool: contract.confirmTool,
        cancelTool: contract.cancelTool,
        compensateTool: contract.compensateTool,
        statusTool: contract.statusTool,
    })) {
        if (companion && companion === contract.toolId) return `${field} refers to the tool itself`;
    }
    return null;
}

function schemaIsObject(text: string): boolean {
    if (!text) return false;
    try {
        const parsed: unknown = JSON.parse(text);
        return typeof parsed === 'object' && parsed !== null && !Array.isArray(parsed);
    } catch {
        return false;
    }
}
